### PR TITLE
Migrate from maven to SBT

### LINKED ISSUE #549: [FEATURE] Evaluate switching build systems
Maven causes us a fair bit of pain (from the slow, non-incremental builds, to the noisy output, to the verbose and unchecked XML config files). We might consider switching to a more advanced build system. Options to evaluate:

- SBT: standard in scala community, highly configurable, incremental builds
- Gradel: noted for being fast, support native compilation, incremental builds, build caches
- Bazel: multi-language, blazingly fast, incremental and cloud builds, could replace all of our use of make (including the benchmarks)

Each has tradeoffs, so in evaluating whether we should switch we should also evaluate which to switch to.

### PR DESCRIPTION
**Pull request by shonfeder**
Closes #549

Migrate the build system from maven to SBT.

## Notes

We used to "ratchet" runs of `scalafmt` with the `spotless` plugin. This had the effect of only running scalafmt on the **parts of files** that had been changed relative to the trunk. However, spotless doesn't support this functionality for sbt. Instead, this change includes a configuration setting to the `sbt-scalafmt` that only run its checks/fixes against **files** that have changed relative to the trunk. This will produce slightly noiser diffs after formatting has been run until all files have been reformatted.

## Principle changes:

- [x] Migrate package management to SBT (see `project/Dependencies.sbt`)
- [x] Migrate multi-package build configuration to SBT (see `build.sbt`)
- [x] Migrate unit tests to SBT (see `build.sbt`)
- [x] Migrate packaging to SBT
- [x] Migrate CI and docker build to SBT
- [x] Migrate release process to SBT (See #1249)
- [x] Configure (ratchted) scalafmt

- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [x] Documentation added for any new functionality
- [x] Entry added to [UNRELEASED.md](./UNRELEASED.md) for any new functionality

### CODE REVIEW

**Review by konnov — APPROVED** [review]
Great! This looks cleaner than the maven files, though I have to get used to the fact that it is actually executable code %)

File: build.sbt [hunk 1]
```diff
@@ -0,0 +1,130 @@
+import Dependencies._
+
+///////////////////////////
+// Project-wide settings //
+///////////////////////////
+
+// See https://www.scala-sbt.org/1.x/docs/Multi-Project.html#Build-wide+settings
+
+ThisBuild / organization := "at.forsyte"
```
[the remark this answers is not in the document]
Comment (shonfeder) [reply build.sbt:9]: My understanding is that this is the package prefix. But I'll check, and If I'm wrong I'll change it :)

File: build.sbt [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+      libraryDependencies ++= Seq(
+          Deps.ujson,
+          Deps.kiama,
+      ),
+  )
+
+lazy val infra = (project in file("mod-infra"))
+  .dependsOn(tlair)
+
+// TODO: Why is this project prefixed with "tla"?
```
[the remark this answers is not in the document]
Comment (shonfeder) [reply build.sbt:73]: Ah, that makes sense. I'll remove the TODO for now.

**Comment by shonfeder** [comment]
Thanks for the initial review! I should opened this as a draft, since I'm still using the pr to debug the CI. :)

**Review by shonfeder — COMMENTED** [review]
Just leaving a few notes to inform review.

**Comment by shonfeder** [comment]
We have code coverage now :) :point_up:

**Comment by shonfeder** [comment]
Contrary to the plan discussed in our standup on Friday, I ended up hacking in integration with our current release pipeline after all. It just didn't feel right to break our release process.

Sorry that this made the diff a bit longer, but I think there are really only a hundred lines or so of logic that needs any consideration.

Please be encouraged to ask any questions about the build configuration, as this can help ensure we all know how to maintain the project going forward, and could point the way to improved dev docs in this PR.

It would also be great if folks could confirm that the build and tests etc. all work as expected on their local systems, so we can catch any system-specific problems before you are locked in to having do you with the new configuration.

**Review by gabrielamafra — APPROVED** [review]
Looks good and works fine on my setup! This is amazing, thank you @shonfeder :heart:

**Comment by konnov** [comment]
A small feature request. Can we test for the presence of `sbt` in the makefile and tell the users what to do, if it is not found? Just to make sure that this is part of the process, not a random bug in the makefile.

**Comment by shonfeder** [comment]
@konnov I think we've so far taken the view that people trying to build our source from scratch should have some minimal technical abilities. If they try to run the makefile without sbt installed, it will tell them `sbt` cannot be found, I think anyone with experience building anything will know that means they need to install sbt. We also have this documented in the user manual. Do you think this isn't enough?

**Comment by konnov** [comment]
Perhaps, it is. I just worry that not so many people have an idea about what `sbt` is about

**Review by konnov — APPROVED** [review]
I tried my standard commands, and they are working. A really great PR. Amazing parallel builds and unit tests!

**Comment by shonfeder** [comment]
Thanks all for the review!

I'm not opposed to adding a guardrail to the makefile re: sbt, but let's consider that for followup work, since that isn't needed to achieve parity with the maven set up.

**Pull request merged**